finality-aleph/src/aggregator.rs

```diff
@@ -82,28 +83,59 @@ impl<
             last_hash_placed: false,
             started_hashes: HashSet::new(),
             metrics,
+            marker: PhantomData,
         }
     }
 
-    pub(crate) async fn start_aggregation(&mut self, hash: B::Hash) {
+    pub(crate) async fn start_aggregation(&mut self, hash: H) {
         debug!(target: "aleph-party", "Started aggregation for block hash {:?}", hash);
         if !self.started_hashes.insert(hash) {
+            debug!(target: "afa", "Aggregation already started for block hash {:?}, exiting.", hash);
             return;
         }
         if let Some(metrics) = &self.metrics {
             metrics.report_block(hash, std::time::Instant::now(), Checkpoint::Aggregating);
         }
         self.hash_queue.push_back(hash);
-        self.rmc.start_rmc(SignableHash { hash }).await;
+        self.rmc.start_multicast(SignableHash { hash }).await;
     }
 
     pub(crate) fn notify_last_hash(&mut self) {
         self.last_hash_placed = true;
     }
 
-    pub(crate) async fn next_multisigned_hash(
-        &mut self,
-    ) -> Option<(B::Hash, MK::PartialMultisignature)> {
+    pub(crate) async fn process_network_messages(&mut self) -> NetworkResult {
```

Comment:
Very nice suggestion, moving the loop inside the function really does make the whole thing clearer. I wanted to extract as much logic as possible out of the loops to make the testing easier but I guess in this case it would mean sacrificing code quality :)